### PR TITLE
feat(assign): unified multi-class picker across Quiz/VA/GL (Phase 5A)

### PR DESCRIPTION
**Pull request by OPS-PIvers**
## Summary

Replaces the split single-select ClassLink dropdown + separate period checkboxes with one shared `AssignClassPicker` that offers a source toggle (ClassLink classes XOR local rosters) and multi-select from the active list. Teachers with multiple ClassLink classes can now target more than one in a single assignment; Video Activity and Guided Learning pick up multi-class targeting plus a post-PIN class-period picker at parity with Quiz.

- **New**: `components/common/AssignClassPicker.{tsx,helpers.ts}` — shared segmented-toggle multi-select picker.
- **Quiz/VA/GL Managers + Widgets**: adopt the picker; `onAssign` now passes `classIds[]` + `selectedPeriodNames[]`. Quiz PLC mode derives period labels from ClassLink class titles when source = ClassLink so PLC export still works.
- **Hooks** (`useQuizAssignments`, `useVideoActivitySession`, `useVideoActivityAssignments`, `useGuidedLearningSession`): write multi-class `classIds` and mirror `classIds[0]` to the legacy `classId` field for one transitional release.
- **Types**: `QuizSession` / `VideoActivitySession` / `GuidedLearningSession` gain `classIds?: string[]`; VA + GL sessions gain `periodNames?: string[]`; VA + GL responses gain `classPeriod?: string`.
- **Student apps**: VA and GL gain the post-PIN class-period picker step mirroring the existing Quiz pattern; picked period is persisted on the response doc.
- **Firestore rules**: new `passesStudentClassGateCompat(classIds, classId)` helper selects the list gate when `classIds` is populated and falls back to the legacy single-class gate otherwise, preserving backward compatibility for sessions created pre-Phase-5A. Applied to `quiz_sessions`, `video_activity_sessions`, and `guided_learning_sessions` (both the session `get` rule and every response-subcollection read/create/update gate).

## Test plan

- [x] `pnpm run validate` — type-check + lint + format + unit tests (1423 passed) + functions tests (143 passed)
- [ ] `pnpm run test:rules` (CI) — verify the existing `studentRoleClassGate` suite still passes with the new compat helper (seeds use legacy `classId` only, which exercises the fallback branch)
- [ ] Manual on dev preview (per verification plan in `/root/.claude/plans/yes-lazy-spindle.md`):
  - [ ] Teacher with 4 ClassLink classes → Quiz Assign → select all 4 → session doc has `classIds: [4]` AND transitional `classId: <first>`
  - [ ] Teacher selects 2 local rosters → session doc has `periodNames: [2]`
  - [ ] ClassLink-SSO student (one of the 4 classes) sees the assignment on `/my-assignments`
  - [ ] PIN-joining student sees post-PIN period picker for multi-period sessions; `classPeriod` is recorded on the response
  - [ ] Same checks for Video Activity and Guided Learning
  - [ ] Quiz PLC mode + ClassLink source: PLC export rows reflect ClassLink class labels
  - [ ] Pre-Phase-5A assignment (legacy `classId` only) still readable by ClassLink-SSO students after rules deploy

## Backward compatibility

- Session writes dual-write `classIds` + `classId = classIds[0]` for one release.
- Rule compat helper preserves single-class gating for sessions with only `classId` (no `classIds` field).
- Per-widget pre-select maps (`lastClassIdBy{Quiz,Activity,Set}Id`) are retained alongside new `lastClassIdsBy{Quiz,Activity,Set}Id` maps so existing widget configs don't lose their memory on first render.

## Out of scope

Listed in the plan file: mini-app index PR (B1), mini-app listener gating (B2), listener-ghost diagnostic (B3), Quiz creation account-state failure (C). Those ship separately.

https://claude.ai/code/session_01Jf5iNcdzDmZjtpBntzFLkL

---
_Generated by [Claude Code](https://claude.ai/code/session_01Jf5iNcdzDmZjtpBntzFLkL)_

### CODE REVIEW

File: components/videoActivity/VideoActivityStudentApp.tsx [hunk 1]
```diff
@@ -96,9 +105,21 @@ const JoinAndPlay: React.FC = () => {
...
-    await joinSession(sessionId, pin.trim(), name.trim());
+    const sessionInfo = await lookupSession(sessionId);
+    const periodNames = sessionInfo?.periodNames ?? [];
+    if (periodNames.length > 1) {
+      setPeriodStep(periodNames);
+      return;
+    }
+    const autoClassPeriod = periodNames[0];
+    await joinSession(sessionId, pin.trim(), name.trim(), autoClassPeriod);
   };
```
[the remark this answers is not in the document]
Comment (OPS-PIvers) [reply components/videoActivity/VideoActivityStudentApp.tsx:127]: Not auto-fixing. This is a UX/product decision (race-condition guard on rapid join clicks) rather than an unambiguous bug — the correct debounce/loading-state strategy depends on the join-flow behavior the team wants (block the button, toast "already joining", spinner overlay, etc.). Flagging for human review.

_Automated reply from the nightly PR-review task._

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by OPS-PIvers — COMMENTED** [review]
## Automated Code Review

**Date:** 2026-04-23

### Summary
Phase 5A of the multi-class assignment picker: introduces a shared `AssignClassPicker` (with source-toggle between ClassLink classes and local rosters) and rolls it out across Quiz / Video Activity / Guided Learning managers and widgets. Sessions now persist `classIds[]` (plus `periodNames[]` where applicable) with a one-release dual-write to the legacy `classId` field, and a new Firestore-rules helper `passesStudentClassGateCompat` selects the list gate when `classIds` is populated and falls back to the single-class gate otherwise. Student-side, VA and GL pick up the post-PIN class-period picker that Quiz already had, and `classPeriod` is persisted on response docs. The scope is cohesive: every collection that writes multi-class updates its types, hook, rules, and student app in lockstep.

### Code Quality
✓ No obvious issues. The dual-write pattern (`classIds` + transitional `classId = classIds[0]`) is the right shape for a one-release compat window. Shared picker lives in `components/common/` which is the right home. Types in `types.ts` pick up `classIds?: string[]` / `periodNames?: string[]` on three session shapes and `classPeriod?: string` on two response shapes — all additive, no breaking narrowings spotted.

### Widget Standards
✓ Not applicable — no new widgets; this PR touches only the manager/settings panels of three existing assignment widgets.

### Admin Config Standards
✓ Not applicable — no admin-settings-level changes.

### Firestore & Security
✓ The `passesStudentClassGateCompat(classIds, classId)` helper is the correct pattern for a backward-compat rules migration: old sessions (legacy `classId` only, no `classIds` field) exercise the fallback branch and continue to gate correctly; new sessions use the list gate. The helper is applied both to the session `get` rule and to every response-subcollection read/create/update gate across `quiz_sessions`, `video_activity_sessions`, and `guided_learning_sessions`, which matches the write-path coverage needed to keep ClassLink-SSO students out of out-of-class submissions. ⚠️ One gap worth a human review: the PR's test plan marks `pnpm test:rules` as a pending CI item rather than confirming it passed locally — the compat helper's fallback branch is only exercised by legacy-shaped seeds (single `classId` field), so make sure the existing `studentRoleClassGate` suite actually seeds sessions without `classIds` to hit that branch. An added test with a seeded session carrying both legacy `classId` and the new `classIds` would also lock in the priority (list gate wins when both are present).

### Regression Risk
⚠️ Moderate. Surface area is large (15 files: 5 components/widgets/ paths, 4 hooks/, rules, types, 3 UI apps). Things to watch in review:
- **`types.ts`**: `classIds?: string[]` is additive, but verify the `ConfigForWidget` helper (if any) doesn't need a new branch, and check that the Quiz/VA/GL `*Session` type consumers that destructure `classId` still compile against the dual-write shape.
- **Session hooks (`useQuizAssignments`, `useVideoActivitySession`, `useVideoActivityAssignments`, `useGuidedLearningSession`)**: these write `classIds` AND mirror `classIds[0]` to `classId`. If `classIds` is ever empty (`.filter(Boolean).length === 0`), `classIds[0]` becomes `undefined` and the mirrored `classId` write needs to handle that — confirm the manager UI blocks submission when the picker has zero selections.
- **Quiz PLC mode**: the description notes the ClassLink-source branch derives period labels from class titles so PLC export still works. This is a specific behavior that's only exercised when `source === 'classlink'` AND PLC mode is toggled — worth an integration test rather than manual-only verification.
- **Per-widget pre-select maps**: the new `lastClassIdsBy{Quiz,Activity,Set}Id` maps live alongside the legacy `lastClassIdBy...` maps. After the one-release transition, a follow-up PR should drop the legacy maps.

### Missing Tests
⚠️ The unit/functions suite (1423 + 143) stays green, but the meaningful new behavior — multi-class targeting, picker source-toggle, post-PIN class-period selection on VA/GL, and the rules compat helper — has no new automated coverage. The description's manual verification checklist is reasonable but all six items are unchecked. At minimum consider: (a) a rules emulator test for `passesStudentClassGateCompat` covering three cases (new `classIds`-only session, legacy `classId`-only session, both present), and (b) a hook-level test for the `classIds` → `classId` mirror write path in the four session hooks.

### Overall
**Merge readiness:** Ready with minor notes
Well-scoped Phase 5A implementation with sensible backward compatibility and a sensible one-release transition plan. The unresolved Gemini comment on `VideoActivityStudentApp.handleJoin` (race-condition guard) is a UX/product decision, not a blocker. The main asks before merge: (1) actually run `pnpm test:rules` and confirm the compat fallback path; (2) check the `classIds[0] === undefined` edge case in each session hook; (3) decide whether the multi-class / period-picker behaviors need new automated coverage before the manual verification checklist can be re-run reliably on dev preview.

---
_This review was generated automatically. All findings should be verified by a human before merging._

---
_Generated by [Claude Code](https://claude.ai/code)_

File: firestore.rules [hunk 2]
```diff
@@ -810,13 +851,19 @@ service cloud.firestore {
...
+      // per-doc class gating still happens at `get`.
       // studentRole (GIS) users additionally must have at least one of the
       // session's classIds in their classIds claim (or the session must be
       // untargeted — empty classIds — for shareable-link launches).
-      allow read: if request.auth != null &&
+      allow get: if request.auth != null &&
         (!isStudentRoleUser() ||
          passesStudentClassGateList(resource.data.get('classIds', [])));
+      allow list: if request.auth != null;
       // Only authenticated (non-anonymous) teachers can create sessions they own
```
[the remark this answers is not in the document]
Comment (OPS-PIvers) [reply firestore.rules:865]: False positive — this rule change is not introduced by the Phase 5A commit. It came from PR #1390 (`4ec4c40 fix(rules): split session reads into get+list to unbreak teacher assignment creation`), which landed on this branch before my commit but hasn't yet merged to `main`. `git show 0c872d1 -- firestore.rules | grep "^[+-].*allow (get|list|read):"` returns no matches — my Phase 5A commit only adds the `passesStudentClassGateCompat` helper and swaps call sites.

The `mini_app_sessions` (and `activity_wall_sessions`) `read` → `get` + `list` splits you see in the diff will land with PR #1390's eventual merge, either via this PR (since #1390 is in this branch's history) or separately if #1390 merges first. No scope expansion intended on my part.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: firestore.rules [hunk 3]
```diff
@@ -969,13 +1025,16 @@ service cloud.firestore {
...
+      // Reads split into `get` + `list`; see quiz_sessions for the rationale
+      // (Firestore static-analyzes `list` predicates and rejects any rule
+      // that touches `resource.data`, even behind a short-circuit OR).
+      // The `list` rule allows any authenticated caller so studentRole users
+      // can run their MyAssignmentsPage class-filtered discovery queries;
+      // per-doc class gating still happens at `get`.
+      allow get: if request.auth != null &&
         (!isStudentRoleUser() ||
          passesStudentClassGate(resource.data.get('classId', '')));
+      allow list: if request.auth != null;
```
[the remark this answers is not in the document]
Comment (OPS-PIvers) [reply firestore.rules]: Same as the `mini_app_sessions` comment above — this `read` → `get` + `list` split came from PR #1390 (commit `4ec4c40`), not from this PR. `git show 0c872d1 -- firestore.rules` confirms the Phase 5A commit makes zero `allow read/get/list` line changes. #1390 is already in this branch's history; the split will land when either PR merges.

---
_Generated by [Claude Code](https://claude.ai/code)_
